bundles/org.eclipse.text/src/org/eclipse/jface/text/source/AnnotationModel.java

```diff
@@ -640,13 +640,23 @@ private void cleanup(boolean fireModelChanged, boolean forkNotification) {
 			fDocumentChanged= false;
 
 			ArrayList<Annotation> deleted= new ArrayList<>();
-			Iterator<Annotation> e= getAnnotationMap().keySetIterator();
 			IAnnotationMap annotations= getAnnotationMap();
-			while (e.hasNext()) {
-				Annotation a= e.next();
-				Position p= annotations.get(a);
-				if (p == null || p.isDeleted())
-					deleted.add(a);
+			
+			if (!(annotations instanceof IAnnotationMap2)) {
```

Comment:
Nitpick: Please change the order of the if and else branch and avoid the negation in the condition. 

Would this work, too?

```java
	private void cleanup(boolean fireModelChanged, boolean forkNotification) {
		if (fDocumentChanged) {
			fDocumentChanged= false;

			ArrayList<Annotation> deleted= new ArrayList<>();
			Object lock = getLockObject();
			if (lock != null) {
				synchronized(lock) {
					getAnnotationMap().forEach(..);
				}
			} else {
			Iterator<Annotation> e= getAnnotationMap().keySetIterator();
			IAnnotationMap annotations= getAnnotationMap();
			while (e.hasNext()) {
				Annotation a= e.next();
				Position p= annotations.get(a);
				if (p == null || p.isDeleted())
					deleted.add(a);
			}
..
	}
```

Reply:
> Would this work, too?

You mean, don't use new interface, or use different lock, or both?
The problem I see is that the map (via ISynchronizable) and model each declare lock objects in their interfaces and do not state that model lock should be same as the map one. So locking on model lock would not help if the "other code" uses map locks to modify the map at same time.

Reply:
> Nitpick: Please change the order of the if and else branch

I wanted to minimize diff, so it is easier to see how code evolved.

Reply:
I think this might be safe to change since both the AnnotationMap and AnnotationModel belong to the same bundle.